Added *.d.ts files in es + update version since npx would allow 3.0-alpha

ralaska opened the pull request.

openbibleinfo commented:
@ralaska I want an npm error, don't I, since I don't want this to end up on npm yet? Or is this error preventing something else? (Admittedly, I didn't realize it would cause an npm error.)

I have a question about the d.ts files. I'm not sure these are useful to a developer since they're very thoroughly specified, and you won't be interacting with these specific files; you'll just be passing them through to the bcv_parser constructor. Are they actually useful? If so, can they be trimmed down so they're less specific? I think specifying bcv_parser.d.ts, where the public API lives, is more important?

ralaska commented:
In Angular, `.d.ts` files are crucial for providing type information. I am using angular @ https://bible.peplamb.com. `.d.ts` files (TypeScript declaration files) are required for `npm install` in TypeScript projects to provide type definitions for JavaScript libraries. These files enhance the developer experience by enabling TypeScript's type-checking and IntelliSense for JavaScript libraries.

ralaska commented:
They were generated using the following command

```sh
npx -p typescript tsc path-to-dir/*.js --declaration --allowJs --emitDeclarationOnly --outDir types
```

Then copied manually to where they need to be (right where the *.js files live)

ralaska commented:
Currently I am using this package for both as `API` and also as `Web Workers` in Angular 19 (https://bible.peplamb.com/). It is on older version right now, thank for accepting my PR for Farsi, Farsi is currently working locally after I added `.d.ts` and some additional changes since we are on `es` now. I will do additional testing on Farsi and will keep you posted.

openbibleinfo commented:
My main concern with the .d.ts files is their high specificity. Since each file exports three objects that you never interact with as an end user, can I just export them as unknown?

So something like the following (I'm not sure what the exact syntax is):

```typescript
export grammar: unknown;
export regexps: unknown;
export translations: unknown;
```

ralaska commented:
Thanks for the feedback, @openbibleinfo!

The `.d.ts` files in the PR were auto-generated using:

```sh
npx -p typescript tsc path-to-dir/*.js --declaration --allowJs --emitDeclarationOnly --outDir types
```

This generates type definitions based on the actual structure of the exports. To simplify them as `unknown`, I’d need to adjust the `.d.ts` files manually or modify them post-generation.

However, introducing manual steps could lead to a maintenance overhead, especially as the project evolves. If this simplification is essential, we could explore automating the process with a script or alternative tooling to ensure consistency.

I was also considering the overall approach for `.d.ts` files and wanted your thoughts:

1. **Keep them in the project**: Makes it easier for TypeScript users (like Angular developers) since the type definitions are bundled directly with the library. However, it adds maintenance overhead and might not benefit most users.

2. **Move them to a separate repository**: Keeps the main project focused and lighter, with type definitions provided as a separate package (e.g., `@types/bible-passage-reference-parser`). This requires coordination between repos but simplifies the core library.

Would you prefer automating the simplification of the `.d.ts` files in the PR using the command above, or exploring a separate repo for types to keep things maintainable? Happy to adjust based on what works best for the project.

openbibleinfo commented:
For the .d.ts files, I think it makes sense to put them directly in the repo. I can easily generate them as part of the build process if you're able to confirm what they should contain. Does the above "unknown" syntax work if you temporarily update one of your auto-generated files? (Sorry, I'm not familiar enough with Typescript definitions to know exactly how they interact with Angular.)

For the book names, I thought I got all of the ones that in your request; which ones do you find aren't working?

ralaska commented:
Thanks for considering adding the `.d.ts` files directly into the repo and incorporating them into the build process. The "unknown" syntax seems like a practical solution to simplify the definitions while keeping them flexible. It should work fine with Angular, as Angular primarily relies on the presence of type definitions rather than their specificity. If you'd like, I can assist in refining the `.d.ts` files to align with this approach for the build process. Let me know how you'd prefer to proceed and yes I can come up with a separate repo specific to `.d.ts`!

Regarding book names please find my tests here:
[Bible_Passage_Reference_Parser_Farsi_AND_Dari_Tests.ipynb](https://github.com/ralaska/Bible-Passage-Reference-Parser-Farsi-Tests/blob/main/Bible_Passage_Reference_Parser_Farsi_AND_Dari_Tests.ipynb)

Dari seems to be working fine (except 3 books failing) weird, I didn't really think about Dari but specifically worked on Farsi. But in Farsi i see that 24 books are failing.

ralaska commented:
I added new tests for Farsi here:

Tests v2:
[Bible_Passage_Reference_Parser_Farsi_AND_Dari_Tests_v2.ipynb](https://github.com/ralaska/Bible-Passage-Reference-Parser-Farsi-Tests/blob/main/Bible_Passage_Reference_Parser_Farsi_AND_Dari_Tests_v2.ipynb)

The PDF i tested wit are using non standard Arabic characters for Farsi.

I also added `convert_to_standard_farsi` which seems to have reduced the errors even further.

As you can see I tested all bible versions' book names in bible.com and the number of books recognized is seems to be contained. If i remove the use of `convert_to_standard_farsi` then the number of books recognized are a bit lesser.

ralaska commented:
Could you please provide step-by-step instructions on how to generate `full.js` and the other files, including details about the machine you’re using? I tried generating them on a fresh setup (both Mac and PC), but it didn’t work for me. Any specific environment settings or tools I should be aware of?

ralaska commented:
@openbibleinfo I updated the `books_names.txt` under `fa` folder based on my tests from:

https://github.com/ralaska/Bible-Passage-Reference-Parser-Farsi-Tests

Would you please review?

openbibleinfo commented:
I created a new release that adds your ~24 missing books to "fa". It looks like your new book_names.txt adds many more. Some of them create conflicts with each other (Pet vs. Pet1), so I didn't add them yet to src/fa/data.txt.

The basic way to create a single language is to run `sh build_lang.sh fa` in /bin (where "fa" can be any language).

To build "full", it's similar, with a caveat:

1. You probably want to run `perl add_cross_lang.pl full` first to update all the book names. As for my environment, I'm running Ubuntu 24.04 with the `sudo apt get libjson-perl` installed. I'm using WSL on Windows.
2. `sh build_lang.sh full`

ralaska commented:
Thank you very much!
